django/utils/functional.py

```diff
@@ -18,13 +21,30 @@ class cached_property:
     Decorator that converts a method with a single self argument into a
     property cached on the instance.
 
-    Optional ``name`` argument allows you to make cached properties of other
-    methods. (e.g.  url = cached_property(get_absolute_url, name='url') )
+    It is possible to make a cached property out of an existing method:
+    (e.g.  ``url = cached_property(get_absolute_url)`` ).
+    On Python versions before 3.6, you'll need to provide the optional ``name``
+    argument. (e.g.  ``url = cached_property(get_absolute_url, name='url')`` ).
     """
     def __init__(self, func, name=None):
+        func_name = func.__name__
+        unsupported_mangled_method = (
+            not PY36 and not name and
+            func_name.startswith('__') and not func_name.endswith('__')
+        )
+        if unsupported_mangled_method:
+            warnings.warn(
+                'cached_property does not work with mangled methods on '
+                'Python 3.5 unless `name` is explicitly set',
+                RuntimeWarning,
+            )
         self.func = func
         self.__doc__ = getattr(func, '__doc__')
-        self.name = name or func.__name__
+        self.name = name or func_name
+
+    def __set_name__(self, owner, name):
+        if self.name == self.func.__name__:
+            self.name = name
 
     def __get__(self, instance, cls=None):
```

Comment:
@graingert `cls` is passed here.

Reply:
I think it's the right class:
```
In [38]: class desc:
    ...:     def __get__(self, instance, cls):
    ...:         if instance is None:
    ...:             return self
    ...:         return instance.__dict__['_%s__mangled' % cls.__name__]

In [39]: class A:
    ...:     d = desc()
    ...: 
    ...:     def __init__(self):
    ...:         self.__mangled = 42
    ...: 

In [40]: A().d
Out[40]: 42
```

Reply:
Try with a subclass of A

Reply:
You might want to add tests for the subclass case @graingert to make it clear.

Reply:
@graingert Probably it doesn't worth it, but still doable :-)
```
In [85]: class desc:
    ...:     def __init__(self, f):
    ...:         self.name = f.__name__
    ...: 
    ...:     def __get__(self, instance, cls):
    ...:         if instance is None:
    ...:             return self
    ...: 
    ...:         for cls in cls.__mro__:
    ...:             name = '_%s%s' % (cls.__name__, self.name)
    ...:             if cls.__dict__.get(name) is self:
    ...:                 return name
    ...: 
    ...: class A:
    ...:     @desc
    ...:     def __x(self):
    ...:         pass
    ...: 
    ...: 
    ...: class B(A):
    ...:     pass
    ...: 

In [86]: a = A()
    ...: a._A__x
    ...: 
Out[86]: '_A__x'

In [87]: b = B()
    ...: b._A__x
    ...: 
Out[87]: '_A__x'
```